### ISSUE TITLE
Implement "exact" bit in list/mapping tagged pointers

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/nballerina#204]
Ballerina makes structures covariant in their member types. This imposes costs on the obvious implementation approach in a couple of ways:

* when a mutable structure is mutated by adding or changing a member, it is necessary to check at runtime whether the member belongs to the type required by the inherent type
* when specialized representations are used for subtypes of a basic type (e.g. representing `byte[]` as an array of bytes), it is usually not possible to infer from the static type what representation is being used (e.g. if you have a variable whose static type is `any[]`, you do not know whether or not it is using the specialized representation that `byte[]` uses or the more general representation); this gets even worse when you have multiple specialized representations (e.g. for `int[]` and for `byte[]` and for tuples).

I believe we can eliminate much of this cost using the following idea. When a variable contains a tagged pointer with uniform type UT_LIST_RW, the tagged pointer can have an additional bit which may be set.  I'll call this the "exact" bit, and it could use either bit 0 or bit 2, (leaving 2 other bits free, assuming  8 byte alignment). When the exact bit is set, it means that the intersection of UT_LIST_RW and the static type of the variable is the same as the inherent type of referenced list. So the code for constructing a list will set the exact bit.  The exact bit will be cleared when there is an implicit upcast (e.g. from byte[] to any[]). This obviously extends to mappings and tables. Note that exactness is a property of the *pointer* to the structure not of the structure itself.

When you have a pointer with the exact bit set, you don't need to do inherent type checking and you know what representation you are dealing with.  You don't know statically whether you will have a pointer to the exact bit, but in performance sensitive code the exact bit is likely to almost always be set. So we would generate code that has an inline, fast path when the exact bit is set, and falls back to a function call when the exact bit isn't set. The exact bit is in the pointer and so LLVM can in principle understand what is going on with it and optimize the exact case even across procedures (although non-integral pointers may interfere with this), particularly if we add appropriate branch weights.

I believe can make this work for members of structures in a similar way: where we use the static type for reference stored in a variable, we use the type that the inherent type requires for a reference stored in a member. If you have an exact reference to a structure `s`, and member `s[k]` is exact, then we can maintain the exact bit when we fetch the member (need to prove this!).

This opens up implementation strategies for structures that would otherwise not be viable.  For example, if we have a record type `record {| int x; int y; |}`, we can represent the fields as `{ i64, i64 }` in LLVM terms; the representation of the record would also have a vtable. When accessed via an exact pointer, member read/write would turn into load/store at an offset (as efficient as C, except for the check for exactness). When accessed via a non-exact pointer, it could go via the vtable.

There's a similar idea that can be used for read-only structures: instead of inherent type, we use the type that was used to choose the representation (need to make this precise).

Final part of this is #572.

**Comment by manuranga** [comment]
I tried to see if llvm is smart enough to hoist such a check out of a loop we naively generate the check per access. [I am not seeing hoisting](https://godbolt.org/z/8ozdsYazv.).

For C++, llvm has [whole program de-vertualizion](https://llvm.org/doxygen/WholeProgramDevirt_8cpp_source.html), but from what I can see it is tightly coupled to C++ type system.

**Comment by jclark** [comment]
I see [fixing your URL](https://godbolt.org/z/8ozdsYazv)

```
sum:                                    # @sum
        mov     rcx, qword ptr [rdi]
        test    rcx, rcx
        jle     .LBB0_1
        xor     edx, edx
        xor     eax, eax
        test    dil, 1
        jne     .LBB0_3
.LBB0_4:                                # =>This Inner Loop Header: Depth=1
        movsx   rsi, byte ptr [rdi + rdx + 16]
        add     rax, rsi
        add     rdx, 1
        cmp     rcx, rdx
        jne     .LBB0_4
        jmp     .LBB0_5
.LBB0_3:                                # =>This Inner Loop Header: Depth=1
        add     rax, qword ptr [rdi + 8*rdx + 16]
        add     rdx, 1
        cmp     rcx, rdx
        jne     .LBB0_3
.LBB0_5:
        ret
.LBB0_1:
        xor     eax, eax
        ret
```

which has two loops, so it *is* hoisting...

**Comment by jclark** [comment]
Exactness can be recovered by doing a cast or by using `is` and implicit type narrowing.

There's some subtlety required to make this work with `readonly`. A type `int[]` means either a mutable array of int, or an immutable array of int. There are a couple of ways to set things up.

One approach is this. A variable with static type `int[]` is exact only if it refers to a mutable array of int. This means if you have a  static type of `int[]`, you can check just for exactness without checking for mutability. With this approach, if you have an immutable array of int, then you would need to say `readonly & int[]` to be exact. Thus doing an implicit upcast from `readonly & int[]` to `int[]` would lose exactness (i.e. clear the exact bit).

Another way to do it would be to say the exact bit is uniform-type specific. So if the uniform type is UT_LIST_RW, then the exact bit means that the inherent type is the static type. In this case, a mutation would require an additional check that the uniform type is mutable (which is just a particular bit in the pointer). If we do this, then we want the representation for UT_LIST_RO and for UT_LIST_RW to be consistent when exact, so we don't have to check on reads.

**Comment by jclark** [comment]
It will work out best, I think, to have the 0x4 bit being set to indicate that the tagged pointer is exact.  The reason to use 0x4 rather than 0x1 is:

- strings use 0x1 for large vs medium strings
- boolean uses 0x1 for true

The reason to use 0x4 rather than 0x2 is that it leaves two contiguous bits (0x1 and 0x2) available for other uses.

Apart from that, the only uses 0f 0x4 are for immediate ints and immediate strings, which will have the IMMEDIATE flag set, after #302 is done. 

The reason to have presence (rather than absence) of the flag meaning exact is that we can use ptrmask to clear the bit when we do a widening conversion.

Another possibility would be to use 0x40 in the top byte i.e. (0x40 << 56).

Or could we reuse the IMMEDIATE flag since no type uses both the IMMEDIATE and EXACT flags? It would making

```
function foo(int[]|int x) returns int?[]|int { return x; }
```

in that we would have to check the tag of the value before we could know whether to clear the exact bit.

**Comment by jclark** [comment]
Suppose p has Ballerina type `int[]`. Then the code to get an element at index i is:

```
#define TAG_MASK (0x1fL << 56)
#define EXACT_MASK 0x2

int _bal_int_list_get(TaggedPtr p, int i) {
  ListPtr p1 = ptrmask(p, TAG_MASK|EXACT_MASK);
  ListPtr p2 = ptrmask(p, TAG_MASK);
  if (likely(p1 != p2)) {
     // exact bit was set: use inlineable fast path
     IntListPtr ilp = (IntListPtr)p1;
     if (unlikely((unsigned)i >= ilp->length)) {
       panic("out of bounds");
     }
     return ilp->members[i];
  }
  else {
     // use slow path through vtable
     return (p1->desc->get)(p, i);
  }
}
```

I am hoping that using ptrmask rather than ptrtoint will enable LLVM to optimize this and hoist it out of loops, even though the tagged pointer is in a non-integral address space.

**Comment by jclark** [comment]
To implement this, we need to identify all cases where there can an implicit widening to a list value. In subset 8, the BIR insns that can do this are:

- ListSet (exactness works for the inherent type of a list member for the static type of a variable)
- MappingSet (similarly)
- ListConstruct
- MappingConstruct
- Call (for the arguments)
- Assign
- Ret

The insns that can restore an exact bit are
- CondNarrow
- TypeCast

In the verifier we should check for exactness rather than subtype for other instructions (we can use `==` since semtypes are anydata).

**Comment by jclark** [comment]
We can implement this before we implement #329.  When we do a ListSet, if the pointer to the list is exact, then (if everything is working) it should be guaranteed that the value being stored is as required by the inherent type of the list. We can verify this and abort if it's not.

**Comment by jclark** [comment]
We can define a _bal_tagged_to_ptr_with_exact function in tagged_inline.ll that takes a TaggedPtr and returns struct with the untagged ptr and a bool saying whether it's exact.

**Comment by jclark** [comment]
We need to be very precise about the semantics of the exact bit.

Every reference to a list has a type constraint associated with it:
- for a BIR register it's the type of the register;
- for a member of a mapping/list, it is the type that the inherent type requires for the member.

Suppose a tagged pointer R has an associated type constraint T and R has the exact bit set. This means that if

- the intersection of T with type UT_LIST_RW is a non-empty type L, and
- R has uniform type UT_LIST_RO or UT_LIST_RW 

then

- R has uniform type UT_LIST_RW, and
- the inherent type of R is L.

So for example: if a variable has type `int[]`, and the value of the variable has the exact bit set, then we know the the value of the variable is a mutable list with inherent type `int[]`.

This means that e.g (neither of these are possible in subset 8):

- widening from `int[]` to `int[]|error` does not clear the exact bit
- widening from `readonly & int[]` to `int[]` does clear the exact bit.

(We will leave more details about the RO case till we implement readonly.)

**Comment by jclark** [comment]
A complication is that this means we can no longer always treat === as being simple pointer comparison, as values that are === may differ in their exact bit.

**Comment by jclark** [comment]
Need to think more about `array:push` (code is currently using erasedSignature which will always widen).

**Comment by jclark** [comment]
Things to do:

- be able to recover the exact bit in CondNarrow and TypeCast
- figure out what to do with `array:push`
- figure out whether ListSet/MappingSet need to behave differently based on the exactness of the reference to the list/mapping

**Comment by jclark** [comment]
We are going to need to distinguish between exact and inexact projections for list/map set/get.

**Comment by jclark** [comment]
For `array:push`, I think we just have to fix `buildCall` to get the last argument of buildWideRepr from funcRef.signature rather than funcRef.erasedSignature.

**Comment by jclark** [comment]
When a structure v has a key k with a member r referring to a structure s, then r having the exact bit set implies that the inherent type of s is the same as the type that the inherent type of v requires for member k. 

So when we do a member get/set, we need to clear the exact bit if
- the type of value being fetched includes mappings/lists, and
- either
   - the reference to the structure does not have the exact bit set, or
   - the key type K is not precise enough

"precise enough" means the projection of T[K] gives the type that the inherent type requires for every k belonging to K.

**Comment by jclark** [comment]
There's now documentation here: https://github.com/ballerina-platform/nballerina/blob/main/docs/exact.md

**Comment by jclark** [comment]
As regards this point:

* figure out whether ListSet/MappingSet need to behave differently based on the exactness of the reference to the list/mapping

the answer is that the exact bit of what is being stored needs to be cleared if the reference to the list/mapping is not exact.

However, this cannot happen in subset 9 because the exact bit will be cleared from any structure value stored in a list/mapping because of widening, except when the inherent type of the structure is `map<any|error>` or `(any|error)[]`. Since these are the top types of the uniform types, if a tagged pointer refers to value with one of these inherent types, it cannot fail to be exact.

**Comment by jclark** [comment]
We will leave nested structure types to a separate issue #572.

All that remains for this issue is https://github.com/ballerina-platform/nballerina/issues/204#issuecomment-939644131
